File: number_app_py/number_app_py/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
from number_app_interfaces.srv import ResetCounter


class NumberCounterNode(Node):
    def __init__(self):
        # Call super()
        super().__init__("number_counter")
        self.counter_ = 0

        # Create publishers, subscribers, services, etc
        self.number_subscriber_ = self.create_subscription(
            Int64, "number", self.callback_number, 10)
        self.reset_counter_service_ = self.create_service(
            ResetCounter, "reset_counter", self.callback_reset_counter)
        self.get_logger().info("Number Counter has been started.")

    # Callback for the subscriber
    def callback_number(self, msg):
        self.counter_ += msg.data
        self.get_logger().info("Counter: " + str(self.counter_))

    # Callback for the service
    def callback_reset_counter(self, request, response):
        if request.reset_value >= 0:
            self.counter_ = request.reset_value
            self.get_logger().info("Counter: " + str(self.counter_))
            response.success = True
        else:
            response.success = False
        return response

def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()

File: number_app_cpp/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "number_app_interfaces/srv/reset_counter.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), counter_(0)
    {
        // Create publishers, subscribers, services, etc
        number_subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10, std::bind(&NumberCounterNode::callbackNumber, this, std::placeholders::_1));
        reset_number_service_ = this->create_service<number_app_interfaces::srv::ResetCounter>(
            "reset_counter",
            std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
        RCLCPP_INFO(this->get_logger(), "Number Counter has been started.");
    }

private:
    // callback for the subscriber
    void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        counter_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "Counter: %d", counter_);
    }

    // callback for the service
    void callbackResetCounter(const number_app_interfaces::srv::ResetCounter::Request::SharedPtr request,
                              const number_app_interfaces::srv::ResetCounter::Response::SharedPtr response)
    {
        if (request->reset_value >= 0)
        {
            counter_ = request->reset_value;
            RCLCPP_INFO(this->get_logger(), "Counter: %d", counter_);
            response->success = true;
        }
        else
        {
            response->success = false;
        }
    }

    // declare class attributes
    int counter_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr number_subscriber_;
    rclcpp::Service<number_app_interfaces::srv::ResetCounter>::SharedPtr reset_number_service_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
